testing/async/test_policy_overloads.h

```diff
@@ -0,0 +1,421 @@
+{
+
+namespace async
+{
+
+// Tests that policies are handled correctly for all overloads of an async
+// algorithm.
+//
+// The AlgoDef parameter type defines an async algorithm, its overloads, and
+// abstracts its invocation. See the async/mixins.h for a documented example of
+// this interface and some convenience mixins that can be used to construct a
+// definition quickly.
+//
+// The AlgoDef interface is used to run several tests of the algorithm,
+// exhaustively testing all overloads for algorithm correctness and proper
+// policy handling.
+//
+// ## Basic tests
+//
+// In the basic tests, each overload is called repeatedly with:
+// 1) No policy
+// 2) thrust::device
+// 3) thrust::device(thrust::device_allocator<void>)
+// 4) thrust::device.on(stream)
+// 5) thrust::device(thrust::device_allocator<void>).on(stream)
+//
+// The output of the async algorithm is compared against a reference output,
+// and the returned event/future is tested to make sure it holds a reference to
+// the expected stream.
+//
+// ## After Future tests
+//
+// The after_future tests check that the future/event returned from an algorithm
+// behaves properly when consumed by a policy's `.after` method.
+template <typename AlgoDef>
+struct test_policy_overloads
+{
+  using algo_def          = AlgoDef;
+  using input_type        = typename algo_def::input_type;
+  using output_type       = typename algo_def::output_type;
+  using postfix_args_type = typename algo_def::postfix_args_type;
+
+  static constexpr std::size_t num_postfix_arg_sets =
+    std::tuple_size<postfix_args_type>::value;
+
+  // Main entry point; call this from a unit test function.
+  static void run(std::size_t num_values)
+  {
+    test_postfix_overloads(num_values);
+  }
+
+private:
+  template <std::size_t size>
+  using size_const = std::integral_constant<std::size_t, size>;
+
+  //----------------------------------------------------------------------------
+  // Recursively call sub tests for each overload set in postfix_args:
+  template <std::size_t postfix_idx = 0>
+  static void test_postfix_overloads(std::size_t const num_values,
+                                     size_const<postfix_idx> = {})
+  {
+    static_assert(postfix_idx < num_postfix_arg_sets, "Internal error.");
+
+    run_basic_policy_tests<postfix_idx>(num_values);
+    run_after_future_test<postfix_idx>(num_values);
+
+    // Recurse to test next round of overloads:
+    test_postfix_overloads(num_values, size_const<postfix_idx + 1>{});
+  }
+
+  static void test_postfix_overloads(std::size_t const,
+                                     size_const<num_postfix_arg_sets>)
+  {
+    // terminal case, no-op
+  }
+
+  //----------------------------------------------------------------------------
+  // For the specified postfix overload set, test the algorithm with several
+  // different policy configurations.
+  template <std::size_t postfix_idx>
+  static void run_basic_policy_tests(std::size_t const num_values)
```

Comment:
Let's just call this one `basic_policy_test` to mirror `after_future_test`.

Reply:
Renamed so that both tests have a `run_XXX_tests` entry point, which can launch multiple `XXX_test` implementations when needed.